cpp/src/barretenberg/honk/proof_system/work_queue.hpp

```diff
@@ -0,0 +1,143 @@
+#include <cstddef>
+#include <memory>
+
+namespace proof_system::honk {
+
+// Currently only one type of work queue operation but there will likely be others related to Sumcheck
+enum WorkType { SCALAR_MULTIPLICATION };
+
+template <typename Params> class work_queue {
+
+    using CommitmentKey = typename Params::CK;
+    using FF = typename Params::Fr;
+    using Commitment = typename Params::C;
+
+    struct work_item_info {
+        uint32_t num_scalar_multiplications;
+    };
+
+    struct work_item {
+        WorkType work_type = SCALAR_MULTIPLICATION;
+        std::span<FF> mul_scalars;
+        std::string label;
+    };
+
+  private:
+    std::shared_ptr<proof_system::plonk::proving_key> key;
+    // TODO(luke): Consider handling all transcript interactions in the prover rather than embedding them in the queue.
+    proof_system::honk::ProverTranscript<FF>& transcript;
+    CommitmentKey commitment_key;
+    std::vector<work_item> work_item_queue;
+
+  public:
+    explicit work_queue(std::shared_ptr<proof_system::plonk::proving_key>& proving_key,
+                        proof_system::honk::ProverTranscript<FF>& prover_transcript)
+        : key(proving_key)
+        , transcript(prover_transcript)
+        , commitment_key(key->circuit_size, "../srs_db/ignition"){}; // TODO(luke): make this properly parameterized
+
+    work_queue(const work_queue& other) = default;
+    work_queue(work_queue&& other) noexcept = default;
+    work_queue& operator=(const work_queue& other) = delete;
+    work_queue& operator=(work_queue&& other) = delete;
+    ~work_queue() = default;
+
+    [[nodiscard]] work_item_info get_queued_work_item_info() const
+    {
+        uint32_t scalar_mul_count = 0;
+        for (const auto& item : work_item_queue) {
+            if (item.work_type == WorkType::SCALAR_MULTIPLICATION) {
+                ++scalar_mul_count;
+            }
+        }
+        return work_item_info{ scalar_mul_count };
+    };
+
+    [[nodiscard]] FF* get_scalar_multiplication_data(size_t work_item_number) const
+    {
+        size_t count = 0;
+        for (const auto& item : work_item_queue) {
+            if (item.work_type == WorkType::SCALAR_MULTIPLICATION) {
+                if (count == work_item_number) {
+                    return const_cast<FF*>(item.mul_scalars.data());
+                }
+                ++count;
+            }
+        }
+        return nullptr;
+    };
+
+    [[nodiscard]] size_t get_scalar_multiplication_size(size_t work_item_number) const
+    {
+        size_t count = 0;
+        for (const auto& item : work_item_queue) {
+            if (item.work_type == WorkType::SCALAR_MULTIPLICATION) {
+                if (count == work_item_number) {
+                    return item.mul_scalars.size();
+                }
+                ++count;
+            }
+        }
+        return 0;
+    };
+
+    void put_scalar_multiplication_data(const Commitment& result, size_t work_item_number)
```

Comment:
plonk passes by value. Are we sure we want a difference here?

Reply:
It's pretty small so probably not a big deal either way but I did this intentionally because all we're doing is serializing the commitment to the transcript. Its scope ends with the function itself so I think no need to pass by value.